rllib/algorithms/appo/appo_learner.py

```diff
@@ -34,69 +44,52 @@ def remove_module(self, module_id: str):
         super().remove_module(module_id)
         self.curr_kl_coeffs_per_module.pop(module_id)
 
-    @override(ImpalaLearner)
-    def additional_update_for_module(
-        self,
-        *,
-        module_id: ModuleID,
-        config: APPOConfig,
-        timestep: int,
-        last_update: int,
-        mean_kl_loss_per_module: dict,
-        **kwargs,
-    ) -> None:
-        """Updates the target networks and KL loss coefficients (per module).
-
-        Args:
-            module_id:
-        """
-
-        #    return dict(
-        #        last_update=self._counters[LAST_TARGET_UPDATE_TS],
-        #        mean_kl_loss_per_module={
-        #            module_id: r[LEARNER_RESULTS_KL_KEY]
-        #            for module_id, r in train_results.items()
-        #            if module_id != ALL_MODULES
-        #        },
-        #    )
-
-        # TODO (avnish) Using steps trained here instead of sampled ... I'm not sure
-        #  why the other implementation uses sampled.
-        #  The difference in steps sampled/trained is pretty
-        #  much always going to be larger than self.config.num_sgd_iter *
-        #  self.config.minibatch_buffer_size unless the number of steps collected
-        #  is really small. The thing is that the default rollout fragment length
-        #  is 50, so the minibatch buffer size * num_sgd_iter is going to be
-        #  have to be 50 to even meet the threshold of having delayed target
-        #  updates.
-        #  We should instead have the target / kl threshold update be based off
-        #  of the train_batch_size * some target update frequency * num_sgd_iter.
-        super().additional_update_for_module(
-            module_id=module_id, config=config, timestep=timestep
-        )
-
-        # TODO (Sven): DQN uses `config.target_network_update_freq`. Can we
-        #  choose a standard here?
-        last_update_ts_key = (module_id, LAST_TARGET_UPDATE_TS)
-        if (
-            timestep - self.metrics.peek(last_update_ts_key, default=0)
-            >= config.target_update_frequency
-        ):
-            self.module._synch_target_network(module_id=module_id, config=config)
-            # Increase lifetime target network update counter by one.
-            self.metrics.log_value((module_id, NUM_TARGET_UPDATES), 1, reduce="sum")
-            # Update the (single-value -> window=1) last updated timestep metric.
-            self.metrics.log_value(last_update_ts_key, timestep, window=1)
-
-        if config.use_kl_loss and module_id in mean_kl_loss_per_module:
-            self._update_module_kl_coeff(
-                module_id, config, mean_kl_loss_per_module[module_id]
-            )
+    @override(Learner)
+    def _after_gradient_based_update(self, *, timesteps: Dict[str, Any]) -> None:
```

Comment:
With this PR, we get rid of `additional_update_for_module` and instead support customizing:

Learner.before_gradient_based_update()
Learner.after_gradient_based_update()

These get called along with the regular `Learner.update()` call, so we won't have the problem of 2x metrics reduction anymore or having to pass around results from the `update()` call back into the `additional_update` call (e.g. the KL values, which felt a little clumsy).

We will still have to streamline this API in the future (maybe make it per module?, give them better names, make them public, unify the `timesteps` arg format).